File: python/data_structures/queue.py
class Node(object):
    def __init__(self, value, prev=None):
        self.value = value
        self.prev = prev


class Queue(object):

    def __init__(self):
        self.clear()

    def isEmpty(self):
        return True if self.tail is None else False

    def enqueue(self, value):
        node = Node(value)
        if self.isEmpty():
            self.tail = node
            self.head = node
        else:
            self.head.prev = node
            self.head = node
        self.size += 1

    def dequeue(self):
        if self.isEmpty():
            raise ValueError("Cannot dequeue on empty stack")
        else:
            if self.head == self.tail:
                self.head = None
            value = self.tail.value
            self.tail = self.tail.prev
            self.size -= 1
            return value

    def clear(self):
        self.tail = None
        self.head = None
        self.size = 0


def main():
    pass


if __name__ == '__main__':
    main()

File: c++/src/data_structures/queue.cpp
#include "data_structures/queue.h"

QueueNode::QueueNode(int value)
{
    value_ = value;
    prev_ = 0;
}

QueueNode::~QueueNode()
{
    value_ = 0;
    prev_ = 0;
}

Queue::Queue()
{
    head_ = 0;
    tail_ = 0;
}

Queue::~Queue()
{
    clear();
}

bool Queue::empty()
{
    return head_ == 0;
}

void Queue::enqueue(int value)
{
    if (head_ == 0) {
        QueueNode* node = new QueueNode(value);
        head_ = node;
        tail_ = node;
    } else {
        QueueNode* node = new QueueNode(value);
        QueueNode* temp = tail_;
        tail_ = node;
        temp->prev_ = tail_;
    }
}

int Queue::dequeue()
{
    if (head_ == 0) {
        throw "Cannot dequeue on empty queue!";
    } else {
        QueueNode* temp = head_;
        int value = head_->value_;
        head_ = head_->prev_;
        if (temp == tail_) {
            tail_ = head_;
        }
        delete temp;
        return value;
    }
}

void Queue::clear()
{
    while (head_ != 0) {
        QueueNode* temp = head_;
        head_ = head_->prev_;
        delete temp;
    }
    tail_ = head_;
}